resources/levels/TileMap.gd

```diff
@@ -0,0 +1,149 @@
+extends TileMap
+
+
+# Damage-Map -> Stores the remaining hardness of cells
+var remaining_hardness_dict = {}
+
+# The layers
+var ground_layer := 0
+var block_layer := 1
+var mine_overlay_layer := 2
+
+# The mine-overlay-atlas
+var mine_overlay_atlas := 1
+# The current's biomes atlassssss
+var block_atlas := 0
+var ground_atlas := 2
+
+# Called when the node enters the scene tree for the first time.
+func _ready():
+	pass # Replace with function body.
+
+
+# Called every frame. 'delta' is the elapsed time since the previous frame.
+func _process(delta):
+	pass
+
+
+# Logic of damaging cells
+func damage_cell(cell: Vector2i, damage: float):
+		var was_ore = false
+		
+		if remaining_hardness_dict.has(cell):
+			# We know the cell -> Further reduce damage
+			remaining_hardness_dict[cell] = remaining_hardness_dict[cell] - damage
+		else:
+			# Unknown cell -> New Entry
+			remaining_hardness_dict[cell] = self.get_cell_tile_data(block_layer, cell).get_custom_data("hardness") - damage
+		self.mine_overlay(cell)
+		
+		var attached_cell = null
+		var cell_was_top :bool
+		# Gotta check if cell itself is a wall
+		if self.get_cell_atlas_coords(block_layer,cell,false).y == 1:
+			attached_cell = self.get_neighbor_cell(cell,TileSet.CELL_NEIGHBOR_TOP_SIDE)
+			cell_was_top = false
+		# Same goes for it being the top of a wall
+		elif self.get_cell_atlas_coords(block_layer,self.get_neighbor_cell(cell,TileSet.CELL_NEIGHBOR_BOTTOM_SIDE),false).y == 1:
+			attached_cell = self.get_neighbor_cell(cell,TileSet.CELL_NEIGHBOR_BOTTOM_SIDE)
+			cell_was_top = true
+		else:
+			cell_was_top = true
+		if attached_cell != null:
+			remaining_hardness_dict[attached_cell] = remaining_hardness_dict[cell]
+			self.mine_overlay(attached_cell)
+		
+		# Remove the tile when it's destroyed
+		if remaining_hardness_dict[cell] < 0:
+			if self.get_cell_tile_data(block_layer, cell).get_custom_data("is_ore"):
+				if cell_was_top:
+					was_ore = cell
+				else: 
+					was_ore = attached_cell
+			self.clear_cell(cell, attached_cell)
+		return was_ore
+
+
+# Puts an overlay on the specified cell
+func mine_overlay(cell: Vector2i) -> void:
+	# Sometimes this function get's called out of sync...
+	if(self.get_cell_tile_data(block_layer, cell) == null):
+			return
+	
+	var percent = 1 - remaining_hardness_dict[cell] / self.get_cell_tile_data(block_layer, cell).get_custom_data("hardness")
+	if percent < 0.25:
+		self.set_cell(mine_overlay_layer,cell,mine_overlay_atlas, Vector2i(0,0), 0)
+		return
+	if percent < 0.5:
+		self.set_cell(mine_overlay_layer,cell,mine_overlay_atlas, Vector2i(1,0), 0)
+		return
+	if percent < 0.75:
+		self.set_cell(mine_overlay_layer,cell,mine_overlay_atlas, Vector2i(2,0), 0)
+		return
+	if percent < 1:
+		self.set_cell(mine_overlay_layer,cell,mine_overlay_atlas, Vector2i(3,0), 0)
+		return
```

Comment:
Might be a bit picky but wouldn't it be cleaner to write else if statements instead of a return in each?

Reply:
fair enough